### PR TITLE
Refactor auth-link state strings to be Base64 encoded EDN maps, and add redirect-origin parameter to allow control over OAuth redirect location

### PR DESCRIPTION
**Pull request by cjsauer**
Trello card: https://trello.com/c/DWVGl7SV/2093-adjust-oauth-state-string-encoding-to-provide-greater-flexibility

Review with: https://github.com/open-company/open-company-web/pull/924

See tests of linked PR.

### CODE REVIEW

**Comment by bago2k4** [comment]
Just FYI (also mine) and for future references the redirect URL needs to be added to distribute apps. We don't need to specify the whole URL, it says it should partially match the URLs listed in the app OAUTH settings:


Also we need to specify the whole redirect URL if we want to use the Install button (that's a static button that we don't control so it needs to have the redirect URL "hardcoded" on their server side.

[Learn more here](https://api.slack.com/docs/oauth)

**Comment by bago2k4** [comment]
@cjsauer so first onboard with Slack went well, then i logged out and logged back in, that worked fine too.
What failed was the bot enable step, to reproduce:
- remove your Carrot org (or completely remove your local DBs)
- onboard with Slack
- open integrations panel
- click on Turn on
- 💥 

Error is:
> Illegal base64 character 2f

I tested this with and without my changes on the web PR: i moved some utility functions out of `oc.web.lib.utils` and into `oc.web.utils.user` since we are trying to gradually reduce the first that is a giant melting pot of shared functions

**Comment by bago2k4** [comment]
also @cjsauer let's deploy this to staging once the above issue is fixed so we have all the team test it

**Comment by bago2k4** [comment]
if you need the stacktrace of the first error
```
19-06-24 13:47:04 bagoPro.local INFO [oc.auth.api.slack:360] - Slack callback
19-06-24 13:47:04 bagoPro.local INFO [oc.auth.lib.slack:108] - Processing Slack response code with Slack state: ezp0ZWFtLWlkICJmZjdjLTQ5MGYtYTk5ZSIsIDp1c2VyLWlkICIyNjRjLTQzYWQtYWM5NiIsIDpyZWRpcmVjdCAiL2NhcnJvdC9hbGwtcG9zdHM/b3JnLXNldHRpbmdzPWludGVncmF0aW9ucyJ9
Mon Jun 24 15:47:04 CEST 2019 [worker-4] ERROR - GET /slack/auth
java.lang.IllegalArgumentException: Illegal base64 character 2f
	at java.util.Base64$Decoder.decode0(Base64.java:714)
	at java.util.Base64$Decoder.decode(Base64.java:526)
	at java.util.Base64$Decoder.decode(Base64.java:549)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at clojure.lang.Reflector.invokeMatchingMethod(Reflector.java:167)
	at clojure.lang.Reflector.invokeInstanceMethod(Reflector.java:102)
	at oc.auth.lib.oauth$decode_state_string.invokeStatic(oauth.clj:14)
	at oc.auth.lib.oauth$decode_state_string.invoke(oauth.clj:11)
	at oc.auth.lib.slack$fixed_decode_state_string.invokeStatic(slack.clj:94)
	at oc.auth.lib.slack$fixed_decode_state_string.invoke(slack.clj:92)
	at oc.auth.lib.slack$swap_code_for_user.invokeStatic(slack.clj:109)
	at oc.auth.lib.slack$swap_code_for_user.invoke(slack.clj:100)
	at oc.auth.lib.slack$oauth_callback.invokeStatic(slack.clj:158)
	at oc.auth.lib.slack$oauth_callback.invoke(slack.clj:151)
	at oc.auth.api.slack$slack_callback.invokeStatic(slack.clj:361)
	at oc.auth.api.slack$slack_callback.invoke(slack.clj:357)
	at oc.auth.api.slack$routes$fn__41479.invoke(slack.clj:395)
	at compojure.core$wrap_response$fn__21224.invoke(core.clj:158)
	at compojure.core$wrap_route_middleware$fn__21208.invoke(core.clj:128)
	at compojure.core$wrap_route_info$fn__21213.invoke(core.clj:137)
	at compojure.core$wrap_route_matches$fn__21217.invoke(core.clj:146)
	at compojure.core$routing$fn__21232.invoke(core.clj:185)
	at clojure.core$some.invokeStatic(core.clj:2701)
	at clojure.core$some.invoke(core.clj:2692)
	at compojure.core$routing.invokeStatic(core.clj:185)
	at compojure.core$routing.doInvoke(core.clj:182)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at compojure.core$routes$fn__21236.invoke(core.clj:192)
	at compojure.core$routing$fn__21232.invoke(core.clj:185)
	at clojure.core$some.invokeStatic(core.clj:2701)
	at clojure.core$some.invoke(core.clj:2692)
	at compojure.core$routing.invokeStatic(core.clj:185)
	at compojure.core$routing.doInvoke(core.clj:182)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at compojure.core$routes$fn__21236.invoke(core.clj:192)
	at raven_clj.ring$wrap_sentry$fn__1528.invoke(ring.clj:22)
	at ring.middleware.params$wrap_params$fn__16361.invoke(params.clj:67)
	at liberator.dev$wrap_trace$fn__16176.invoke(dev.clj:239)
	at liberator.dev$handle_and_add_trace_link.invokeStatic(dev.clj:194)
	at liberator.dev$handle_and_add_trace_link.invoke(dev.clj:193)
	at liberator.dev$wrap_trace_ui$fn__16158.invoke(dev.clj:212)
	at liberator.dev$wrap_trace_header$fn__16163.invoke(dev.clj:216)
	at ring.middleware.cors$allow_origins.invokeStatic(cors.clj:48)
	at ring.middleware.cors$allow_origins.doInvoke(cors.clj:41)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:669)
	at clojure.core$apply.invoke(core.clj:660)
	at ring.middleware.cors$wrap_cors$fn__17046.invoke(cors.clj:59)
	at buddy.auth.middleware$wrap_authentication$fn__18772.invoke(middleware.clj:59)
	at ring.middleware.reload$wrap_reload$fn__17005.invoke(reload.clj:39)
	at org.httpkit.server.HttpHandler.run(RingHandler.java:115)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Review by bago2k4 — DISMISSED** [review]
See comments above

**Comment by belucid** [comment]
Putting these aside for now. Don't need this change at the moment.

**Comment by cjsauer** [comment]
Reopening this PR now that native Google sign-in is an issue.

**Comment by cjsauer** [comment]
@bago2k4 I fixed the Slack bot integration issue that you mentioned above. It was indeed a Base64 decoding problem server-side. Should work now.

File: src/oc/auth/api/google.clj [hunk 1]
```diff
@@ -14,14 +15,14 @@
 
 (defn- redirect-to-web-ui
   "Send them back to a UI page with an access description ('google' or 'failed') and a JWToken."
-  ([redirect access]
+  ([redirect-origin redirect access]
     (redirect-to-web-ui redirect access nil :not-a-new-user)) ; nil = no jwtoken
```
Comment (bago2k4) [inline src/oc/auth/api/google.clj]: @cjsauer i think you are missing a parameter here, shouldn't you pass also redirect-origin in?

**Review by bago2k4 — CHANGES_REQUESTED** [review]
One thing, tests are working though...

File: src/oc/auth/api/google.clj (hunk 1 above)
Comment (cjsauer) [reply src/oc/auth/api/google.clj]: @bago2k4 good catch, fixed.

**Review by bago2k4 — APPROVED** [review]

**Pull request merged**